examples/gno.land/r/gnoland/valopers/valopers.gno

```diff
@@ -0,0 +1,468 @@
+}
+
+// Render renders the current valoper set.
+// "/r/gnoland/valopers" lists all valopers, paginated.
+// "/r/gnoland/valopers:addr" shows the detail for the valoper with the addr.
+func Render(fullPath string) string {
+	path, _ := splitPathAndQuery(fullPath)
+	if path == "" {
+		return renderHome(fullPath)
+	} else {
+		if len(path) < 2 || path[:2] != "g1" {
+			return "invalid address " + path
+		}
+		valoperRaw, exists := valopers.Get(path)
+		if !exists {
+			return "unknown address " + path
+		}
+		v := valoperRaw.(Valoper)
+		return "Valoper's details:\n" + v.Render()
+	}
+}
+
+func renderHome(path string) string {
+	// if there are no valopers, display instructions
+	if valopers.Size() == 0 {
+		return ufmt.Sprintf("%s\n\nNo valopers to display.", instructions)
+	}
+
+	page := pager.NewPager(valopers, 50, false).MustGetPageByPath(path)
+
+	output := ""
+
+	// if we are on the first page, display instructions
+	if page.PageNumber == 1 {
+		output += ufmt.Sprintf("%s\n\n", instructions)
+	}
+
+	for _, item := range page.Items {
+		v := item.Value.(Valoper)
+		output += ufmt.Sprintf(" * [%s](/r/gnoland/valopers:%s) - [profile](/r/demo/profile:u/%s)\n",
+			v.Moniker, v.Address, v.Caller)
+	}
+
+	output += "\n"
+	output += page.Picker()
+	return output
+}
+
+func splitPathAndQuery(fullPath string) (string, string) {
+	parts := strings.SplitN(fullPath, "?", 2)
+	path := parts[0]
+	queryString := ""
+	if len(parts) > 1 {
+		queryString = "?" + parts[1]
+	}
+	return path, queryString
+}
+
+// Validate checks if the fields of the Valoper are valid
+func (v *Valoper) Validate() error {
+	errs := &combinederr.CombinedError{}
+
+	errs.Add(validateMoniker(v.Moniker))
+	errs.Add(validateDescription(v.Description))
+	errs.Add(validateBech32(v.Address.String()))
+	errs.Add(validatePubKey(v.PubKey))
+
+	for _, addr := range v.P2PAddresses {
+		errs.Add(validateP2PAddress(addr))
+	}
+
+	if errs.Size() == 0 {
+		return nil
+	}
+
+	return errs
+}
+
+// Render renders a single valoper with their information
+func (v Valoper) Render() string {
+	output := ufmt.Sprintf("## %s\n", v.Moniker)
+
+	if v.Description != "" {
+		output += ufmt.Sprintf("%s\n\n", v.Description)
+	}
+
+	output += ufmt.Sprintf("- Address: %s\n", v.Address.String())
+	output += ufmt.Sprintf("- PubKey: %s\n", v.PubKey)
+	output += "- P2P Addresses: [\n\n"
+
+	if len(v.P2PAddresses) == 0 {
+		output += "]\n\n"
+	}
+
+	for index, addr := range v.P2PAddresses {
+		output += "\t- " + addr
+
+		if index == len(v.P2PAddresses)-1 {
+			output += "\n\n]\n\n"
+
+			continue
+		}
+
+		output += ",\n"
+	}
+
+	output += ufmt.Sprintf("[Profile link](/r/demo/profile:u/%s)\n", v.Address)
+
+	return output
+}
+
+// isValoper checks if the valoper exists
+func isValoper(address std.Address) bool {
+	_, exists := valopers.Get(address.String())
+
+	return exists
+}
+
+func validateSentCoins() error {
+	sentCoins := std.OriginSend()
+
+	if len(sentCoins) == 1 && sentCoins[0].IsGTE(minCoin) {
```

Comment:
check for "GNOT".

Reply:
`minCoin` is already defined with the denomination "ugnot": `std.NewCoin("ugnot", minFee)`. And `IsGTE` checks that `sentCoins` has the same denomination. I added a test which checks for panic if `sentCoins` is not ugnot:
https://github.com/gnolang/gno/pull/3830/commits/ed8ad8c0629e75a4ceecf4fe62a36dad818a2cd6